### ISSUE TITLE
RPC client returns hbase exception

### ISSUE THREAD

**Issue by ccbeloy** [falsecz/hbase-rpc-client#18]
HBase Version	1.0.0-cdh5.4.5, revision=Unknown
Hadoop Version	2.6.0-cdh5.4.5, revision=ab14c89fe25e9fb3f9de4fb852c21365b7c5608b

java.lang.NullPointerException
	at org.apache.hadoop.hbase.quotas.RegionServerQuotaManager.checkQuota(RegionServerQuotaManager.java:182)
	at org.apache.hadoop.hbase.quotas.RegionServerQuotaManager.checkQuota(RegionServerQuotaManager.java:137)
	at org.apache.hadoop.hbase.regionserver.RSRpcServices.get(RSRpcServices.java:1775)
	at org.apache.hadoop.hbase.protobuf.generated.ClientProtos$ClientService$2.callBlockingMethod(ClientProtos.java:31439)
	at org.apache.hadoop.hbase.ipc.RpcServer.call(RpcServer.java:2035)
	at org.apache.hadoop.hbase.ipc.CallRunner.run(CallRunner.java:107)
	at org.apache.hadoop.hbase.ipc.RpcExecutor.consumerLoop(RpcExecutor.java:130)
	at org.apache.hadoop.hbase.ipc.RpcExecutor$1.run(RpcExecutor.java:107)
	at java.lang.Thread.run(Thread.java:745)

Please advise. Thanks.

**Comment by wision** [comment]
@ccbeloy Can you please post code snippet as well as data sample? Thanks.

**Comment by ccbeloy** [comment]
Please see below code:

describe('Test hbase rpc client v0.1.2:', function() {
    it('should get the row', function (done) {        
        var client = hbase({
            zookeeperHosts: ["10.184.46.248:2181","10.184.46.25:2181","10.184.46.242:2181"],
            zookeeperRoot: "/hbase",
            rootRegionZKPath: "/meta-region-server",
            rpcTimeout: 30000,
            pingTimeout: 30000,
            callTimeout: 5000
        });
        
        client.on("error", function (err) {
            return console.log("hbase client error:", err);
        });
        var get = new hbase.Get('myrowkey');
        
        client.get('testtable', get, function (err, res) {
            console.log('error : ' + JSON.stringify(err));
            console.log('resp : ' + JSON.stringify(res));
            done();
        });
        
        //var del;
        //del = new hbase.Delete("mytestrow");
        //client.delete("default:ooxp" , del, function (err, res) {
        //    return console.log(arguments);
        //});
        assert.ok(true, "This shouldn't fail");
    })
})

**Comment by wision** [comment]
I just tried as close example as you posted and see no problem. I used hbase 1.0.0 (http://archive.apache.org/dist/hbase/hbase-1.0.0/)
```coffeescript
hbase = require("./");

var client = hbase({
        zookeeperHosts: ["localhost:2181"],
        zookeeperRoot: "/hbase",
        rootRegionZKPath: "/meta-region-server",
        rpcTimeout: 30000,
        pingTimeout: 30000,
        callTimeout: 5000
});

client.on("error", function (err) {
        return console.log("hbase client error:", err);
});

var get = new hbase.Get('myrowkey');

client.get('testtable', get, function (err, res) {
        console.log('error : ' + JSON.stringify(err));
        console.log('resp : ' + JSON.stringify(res));
});
```

console output:
```output
15:06:51 ~/sbks/hbase-rpc-client [master]$ node test.js
error : null
resp : {"row":[109,121,114,111,119,107,101,121],"cols":{"cf1:my_column":{"value": 109,121,95,118,97,108,117,101],"timestamp": "low":-2057728161,"high":336,"unsigned":true}}},"columns":[{"family":[99,102,49],"qualifier": 109,121,95,99,111,108,117,109,110],"value":[109,121,95,118,97,108,117,101],"timestamp":{"low":-2057728161,"high":336,"unsigned":true}}]}
```

hbase shell:
```hbase
hbase(main):005:0> scan 'testtable'
ROW                       COLUMN+CELL
 myrowkey                 column=cf1:my_column, timestamp=1445346250591, value=my_value
1 row(s) in 0.0250 seconds
```

Maybe you can get me debug print from console? Run your script with DEBUG=hbase* env variable.

**Comment by Lazyshot** [comment]
I can reproduce as well, it is related to not passing the userInfo field into the connection header. This may be a unique thing to CDH HBase.

Once I add the following snippet:

```js
         ch = new ConnectionHeader({
            serviceName: "ClientService",
            userInfo: new UserInformation({
              effectiveUser: ""
            })
          });
```

This error went away. 

Looks like it's an upcoming addition in 1.1 that is patched into the 1.0 CDH version.
https://issues.apache.org/jira/browse/HBASE-11598

**Comment by ccbeloy** [comment]
Hi @Lazyshot , kindly post full code as I'm not sure how to use ConnectionHeader and which function/object to pass it. Thanks a lot in advance.

**Comment by wision** [comment]
@ccbeloy You will have to modify hbase-rpc-client source code.. something like this:
```diff
11:37:56 ~/sbks/hbase-rpc-client [master]$ git diff
diff --git a/src/connection.coffee b/src/connection.coffee
index 42fc5f4..751a97d 100644
--- a/src/connection.coffee
+++ b/src/connection.coffee
@@ -17,7 +17,7 @@ proto      = builder.build()
 rpcProto   = rpcBuilder.build()
 
 {ClientService, GetRequest} = proto
-{ConnectionHeader, RequestHeader, ResponseHeader} = rpcProto
+{ConnectionHeader, RequestHeader, ResponseHeader, UserInformation} = rpcProto
 
 
 
@@ -54,6 +54,9 @@ module.exports = class Connection extends EventEmitter
                        debug "connected to #{@name}"
                        @writeHead()
                        ch = new ConnectionHeader
+                               userInfo: new UserInformation
+                                       effectiveUser: "someEffectiveUser"
+                                       realUser: "someRealUser"
                                serviceName: "ClientService"
                        header = ch.encode().toBuffer()
                        @out.writeInt header.length
```

**Comment by Lazyshot** [comment]
Yes, that exactly.

**Comment by ccbeloy** [comment]
I'm getting "timedout" on client.get callback

error : "timedout"
resp : undefined

**Comment by wision** [comment]
Please post the output of your script running it with DEBUG=*

**Comment by wision** [comment]
Fixed in #39
